### PR TITLE
Add operator modulo

### PR DESCRIPTION
**Pull request by pmqtt**
Hello folks,
I added the modulo operator to Carbon! A test file showing usage is added in testdatat/operator/operator-modulo.carbon.
I hope it's useful!

Best regrads

### CODE REVIEW

**Comment by pmqtt** [comment]
Has anybody a idea, why the pre-commit workflow not started?

**Comment by pacion** [comment]
What about writing modulo tests with negative numbers?

**Comment by pmqtt** [comment]
> What about writing modulo tests with negative numbers?

I have added tests with negative values... Thanks for looking!

**Review by pacion — APPROVED** [review]

**Comment by pmqtt** [comment]
After failed test, I have check all modulo results in testdata/operator/operator-modulo.carbon with the c++ results

**Comment by pmqtt** [comment]
@pacion can you approve this PR or anybody else?

**Comment by geoffromer** [comment]
> @pacion can you approve this PR or anybody else?

I'm afraid it may take us a few days to review this. Right now we're a bit overwhelmed (but delighted!) with how dramatically our community has grown in the last week, so please bear with us.

**Comment by jonmeow** [comment]
Quick note, it may be helpful to review comments at https://github.com/carbon-language/carbon-lang/issues/1549#issuecomment-1193032785, similar will probably apply here but I haven't looked closely yet.

**Comment by pmqtt** [comment]
> Quick note, it may be helpful to review comments at [#1549 (comment)](https://github.com/carbon-language/carbon-lang/issues/1549#issuecomment-1193032785), similar will probably apply here but I haven't looked closely yet.

Hi, 
what you mean? Implement modulo in this way 
```
 var i : i32 = mod(15,2);
```
?
If not, modulo binds weaker then operator +,- . For this reason, the grammar must be extends to handle the modulo operator.

File: explorer/syntax/parser.ypp [hunk 1]
```diff
@@ -491,7 +513,7 @@ unimpl_expression:
 ;
 value_expression:
   // ref_deref_expression excluded due to precedence diamond.
-  additive_expression
+  modulo_expression
 | as_expression
```
Comment (zygoloid) [inline explorer/syntax/parser.ypp]:
```suggestion
  additive_expression
| modulo_expression
| as_expression
```

Per the [precedence diagram](https://github.com/carbon-language/carbon-lang/tree/trunk/docs/design/expressions#precedence), additive and modulo expressions are at the same precedence level, so it seems simpler to represent them as siblings in the grammar.

File: explorer/syntax/parser.ypp [hunk 2]
```diff
@@ -446,10 +450,28 @@ multiplicative_expression:
...
+  ref_deref_expression
+| additive_expression
+;
+modulo_lhs:
+  ref_deref_expression
+| modulo_expression
+;
+modulo_expression:
+  additive_expression
+| modulo_lhs MODULO modulo_operand
```
Comment (zygoloid) [inline explorer/syntax/parser.ypp]:
```suggestion
modulo_operand:
  ref_deref_expression
| minus_expression
;
modulo_expression:
  modulo_operand MODULO modulo_operand
```

In Carbon, an unparenthesized additive expression isn't permitted as an operand of a modulo expression -- `a % b + c` requires parentheses, for example. Also, modulo expressions are not associative -- `a % b % c` requires parentheses. See the [precedence diagram](https://github.com/carbon-language/carbon-lang/tree/trunk/docs/design/expressions#precedence).

Comment (zygoloid) [reply explorer/syntax/parser.ypp]: Also, I wonder if we can give a better name to the `ref_deref_expression | minus_expression` rule that we now use as both `multiplicative_operand` and `modulo_operand`, and will be used a few more times for other operators too. Maybe `unary_expression`?

File: explorer/testdata/operator/operator-modulo.carbon [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+  if( not (x == 0)){
+      return 2;
+  }
+
+  x = 4+4+4+4 % 2+1;
+  if(  not (x == 1) ){
+      return 3;
+  }
+
+  x = 15 % identFn(2)+1;
```
Comment (zygoloid) [inline explorer/testdata/operator/operator-modulo.carbon]: These three examples are all invalid per the currently-approved design and require parentheses; see [proposal 1083](https://github.com/carbon-language/carbon-lang/blob/trunk/proposals/p1083.md#use-the-same-precedence-group-for-modulo-and-multiplication) for details.

**Review by zygoloid — CHANGES_REQUESTED** [review]

File: explorer/testdata/operator/operator-modulo.carbon (hunk 3 above)
Comment (pmqtt) [reply explorer/testdata/operator/operator-modulo.carbon]: I must admit that I have not seen this proposal. I figured the C++ operator precedence would be fine

File: explorer/syntax/parser.ypp (hunk 2 above)
Comment (pmqtt) [reply explorer/syntax/parser.ypp]: Yes, I think you are reigh!t The best name is unary_expression...

**Comment by pmqtt** [comment]
@zygoloid I implemented your review and corrected the test cases. Many thanks

**Review by zygoloid — COMMENTED** [review]
Thanks, this looks good to me, but the patch has merge conflicts with the change to support overloading for  `+`, `-`, `*`. Please can you resolve those conflicts?

It would be great if you also added overloading support for `%`, but I'm happy for this to go ahead without overloading support if you prefer.

**Comment by pmqtt** [comment]
@zygoloid Merge conflict is resolved...

File: explorer/syntax/parser.ypp [hunk 4]
```diff
@@ -446,10 +449,23 @@ multiplicative_expression:
...
+      $$ = arena->New<OperatorExpression>(
+          context.source_loc(), Operator::Mod,
+          std::vector<Nonnull<Expression*>>({$1, $3}));
+    }
+;
 additive_operand:
   ref_deref_expression
 | multiplicative_expression
 ;
+
```
Comment (zygoloid) [inline explorer/syntax/parser.ypp]:
```suggestion
```
Let's keep the `additive_` productions together.

File: explorer/testdata/operator/operator-modulo.carbon [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+  }
+
+  x = -15 % 7;
+  if( not (x == -1)){
+      return 5;
+  }
+
+  x = 15 % -7;
+  if( not ( x == 1)){
+      return x;
```
Comment (zygoloid) [inline explorer/testdata/operator/operator-modulo.carbon]:
```suggestion
      return 6;
```
Just avoiding a false-negative if x is somehow 0 here.

**Review by zygoloid — APPROVED** [review]

**Pull request merged**